modules/ref_fb_module/src/struct_decoder_fb_impl.cpp

```diff
@@ -0,0 +1,289 @@
+namespace StructDecoder
+{
+
+static const char* InputDisconnected = "Disconnected";
+static const char* InputConnected = "Connected";
+static const char* InputInvalid = "Invalid";
+
+StructDecoderFbImpl::StructDecoderFbImpl(const ContextPtr& ctx, const ComponentPtr& parent, const StringPtr& localId)
+    : FunctionBlock(CreateType(), ctx, parent, localId)
+{
+    createInputPorts();
+    initStatuses();
+}
+
+FunctionBlockTypePtr StructDecoderFbImpl::CreateType()
+{
+    return FunctionBlockType("RefFBModuleStructDecoder", "Struct decoder", "Decodes signals with struct data type and outputs signal for each struct component.");
+}
+
+void StructDecoderFbImpl::processSignalDescriptorChanged(const DataDescriptorPtr& inputDataDescriptor,
+                                                         const DataDescriptorPtr& inputDomainDataDescriptor)
+{
+    if (inputDataDescriptor.assigned())
+        this->inputDataDescriptor = inputDataDescriptor;
+    if (inputDomainDataDescriptor.assigned())
+        this->inputDomainDataDescriptor = inputDomainDataDescriptor;
+
+    configure();
+}
+
+void StructDecoderFbImpl::configure()
+{
+    if (!inputDataDescriptor.assigned() || !inputDomainDataDescriptor.assigned())
+    {
+        setInputStatus(InputInvalid);
+        return;
+    }
+
+    try
+    {
+        if (inputDataDescriptor.getDimensions().getCount() > 0)
+            throw std::runtime_error("Arrays not supported");
+
+        const auto inputSampleType = inputDataDescriptor.getSampleType();
+        if (inputSampleType != SampleType::Struct)
+            throw std::runtime_error("Invalid sample type");
+
+        signals.clear();
+
+        constexpr std::array<SampleType, 11> validFieldTypes{
+            SampleType::Int8,
+            SampleType::Int16,
+            SampleType::Int32,
+            SampleType::Int64,
+            SampleType::UInt8,
+            SampleType::UInt16,
+            SampleType::UInt32,
+            SampleType::UInt64,
+            SampleType::Float32,
+            SampleType::Float64,
+            SampleType::Struct,
+        };
+
+        const auto domainSignal = createAndAddSignal("__domain", inputDomainDataDescriptor);
+        domainSignal.asPtr<IComponentPrivate>().unlockAttributes(List<IString>("Visible"));
+        domainSignal.setVisible(false);
+        domainSignal.asPtr<IComponentPrivate>().lockAttributes(List<IString>("Visible"));
+
+        const auto structFields = inputDataDescriptor.getStructFields();
+        for (const auto& field: structFields)
+        {
+            const auto fieldSampleType = field.getSampleType();
+            if (std::find(validFieldTypes.begin(), validFieldTypes.end(), fieldSampleType) == validFieldTypes.end())
+                throw std::runtime_error(fmt::format("Field \"{}\" has invalid sample type", field.getName()));
+
+            const auto signal = createAndAddSignal(field.getName(), field);
+            signal.setDomainSignal(domainSignal);
+        }
```

Comment:
This should probably be recursive for nested struct-type signals, unless this fb should only decode 1 layer of the structs. Similar comments apply to the data processing.

Reply:
It should decode just one layer and then additional decoder would be needed. I want to keep the implementation simple. Structs are anyway rare and nested structs even more :)